### PR TITLE
feat(dop): Delete unnecessary keys in project pipeline configuration

### PR DESCRIPTION
**Pull request by kakj-go**
#### What this PR does / why we need it:
Delete unnecessary keys in project pipeline configuration

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMTY0XSwiYXNzaWduZWUiOlsiMTAwMDU2MCJdfQ%3D%3D&id=299271&iterationID=1164&pId=293275&type=TASK)

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      Delete unnecessary keys in project pipeline configuration        |
| 🇨🇳 中文    |        项目流水线配置删除不需要的key      |

### CODE REVIEW

File: modules/dop/providers/cms/cicdcmsservice.go [hunk 1]
```diff
@@ -127,6 +131,48 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
...
+			Ns:             updateRequest.Ns,
+		})
+		if err != nil {
+			return false, err
+		}
+
+		// get unnecessary keys
+		var deleteKeys []string
+		for _, v := range preCMS.Data {
+			if v.Type == "FILE" {
```
Comment (sfwn) [inline modules/dop/providers/cms/cicdcmsservice.go:148]: file type config also can be deleted

File: modules/dop/providers/cms/cicdcmsservice.go [hunk 2]
```diff
@@ -127,6 +131,48 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
...
+				continue
+			}
+
+			var find = false
+			for _, conf := range keys {
+				if conf != v.Key {
+					continue
+				}
+				find = true
```
Comment (sfwn) [inline modules/dop/providers/cms/cicdcmsservice.go:157]: break

File: modules/dop/providers/cms/cicdcmsservice.go [hunk 3]
```diff
@@ -127,6 +131,48 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
...
+			Ns:             updateRequest.Ns,
+			PipelineSource: updateRequest.PipelineSource,
+			DeleteKeys:     deleteKeys,
+		})
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// create or update configs
```
Comment (sfwn) [inline modules/dop/providers/cms/cicdcmsservice.go:175]: what will happen if delete success but update failed?

File: modules/dop/providers/cms/cicdcmsservice.go [hunk 4]
```diff
@@ -127,6 +131,48 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
 		}))
 	}()
 
+	if req.Batch {
+		// get pre configs
+		newContext := transport.WithHeader(ctx, metadata.New(map[string]string{httputil.InternalHeader: "true"}))
```
Comment (sfwn) [inline modules/dop/providers/cms/cicdcmsservice.go:136]: use apis.WithInternalClientContext

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
Add Unit-Test.
Test it carefully. DO NOT DELETE BY MISTAKE.

File: modules/dop/providers/cms/cicdcmsservice.go (hunk 1 above)
Comment (kakj-go) [reply modules/dop/providers/cms/cicdcmsservice.go]: file type not support edit key

File: modules/dop/providers/cms/cicdcmsservice.go (hunk 3 above)
Comment (kakj-go) [reply modules/dop/providers/cms/cicdcmsservice.go]: The changed key will be lost

**Comment by kakj-go** [comment]
> Add Unit-Test. Test it carefully. DO NOT DELETE BY MISTAKE.

done

File: modules/dop/providers/cms/cicdcmsservice.go [hunk 5]
```diff
@@ -127,13 +127,72 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
...
+	})
+	if err != nil {
+		return err
+	}
+
+	// get unnecessary keys
+	var deleteKeys []string
+	for _, v := range preCMS.Data {
+		// file type not support change key, so can not delete not find key
+		if v.Type == "FILE" {
```
Comment (sfwn) [inline modules/dop/providers/cms/cicdcmsservice.go:161]: i think this type is wrong.
do you test it?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/dop/providers/cms/cicdcmsservice.go (hunk 5 above)
Comment (kakj-go) [reply modules/dop/providers/cms/cicdcmsservice.go]: Changed to dice file type

Comment (kakj-go) [reply modules/dop/providers/cms/cicdcmsservice.go]: It has been tested before, and there is no problem with wrong type judgment

**Review by sfwn — APPROVED** [review]

**Pull request merged**